File: LeetCode/cpp/1019.cpp
class Solution {
public:
    vector<int> nextLargerNodes(ListNode *head) {
        vector<int> s;
        while (head) {
            s.push_back(head->val);
            head = head->next;
        }
        int n = s.size();
        stack<int> temp;
        vector<int> ret(n, 0);
        for (int i = n - 1; i >= 0; --i) {
            while (!temp.empty() && temp.top() <= s[i])
                temp.pop();
            ret[i] = temp.empty() ? 0 : temp.top();
            temp.push(s[i]);
        }
        return ret;
    }
};

File: LeetCode/python3/1019.py
class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        nums = []
        while head:
            nums.append(head.val)
            head = head.next
        n = len(nums)
        ret = [0 for _ in range(n)]
        max_stack = []
        for i in range(n - 1, -1, -1):
            while (len(max_stack) > 0 and max_stack[-1] <= nums[i]):
                max_stack.pop()
            ret[i] = 0 if len(max_stack) == 0 else max_stack[-1]
            max_stack.append(nums[i])
        return ret
